Web Inspector: Timelines: Script: make the target into a hierarchical path component

dcrousso opened the pull request:
#### b8bd34d2de248effbd779dca443b2a583b80837e
<pre>
Web Inspector: Timelines: Script: make the target into a hierarchical path component
<a href="https://bugs.webkit.org/show_bug.cgi?id=288262">https://bugs.webkit.org/show_bug.cgi?id=288262</a>

Reviewed by BJ Burg.

Instead of having a `WI.DataGrid` column for the `target`/&quot;context&quot; in `WI.ScriptDetailsTimelineView` and a dropdown to switch the `WI.CallingContextTree` based on the selected `WI.Target` in `WI.ScriptProfileTimelineView`, the `WI.Target` should be a hierarchical path component to have similar behavior to elsewhere in Web Inspector.

This also means that distinct `WI.ScriptProfileTimelineView` and `WI.ScriptDetailsTimelineView` are created for each `WI.Target` (as needed) instead of having each hold and switch between or display all `WI.Target` in the `WI.TimelineRecording`.

Note that the `WI.HierarchicalPathComponent` for the `WI.Target` is only shown if there&apos;s more than one `WI.Target` in the `WI.TimelineRecording`.

* Source/WebInspectorUI/UserInterface/Views/ScriptClusterTimelineView.js:
(WI.ScriptClusterTimelineView):
(WI.ScriptClusterTimelineView.createPathComponent):
(WI.ScriptClusterTimelineView.prototype.reset):
(WI.ScriptClusterTimelineView.prototype.get selectionPathComponents):
(WI.ScriptClusterTimelineView.prototype.saveToCookie):
(WI.ScriptClusterTimelineView.prototype.restoreFromCookie):
(WI.ScriptClusterTimelineView.prototype.closed):
(WI.ScriptClusterTimelineView.prototype._updateCurrentContentView):
(WI.ScriptClusterTimelineView.prototype._createTargetPathComponent): Added.
(WI.ScriptClusterTimelineView.prototype._sortTargetPathComponents): Added.
(WI.ScriptClusterTimelineView.prototype._handleViewPathComponentSelected):
(WI.ScriptClusterTimelineView.prototype._handleTargetAdded): Added.
(WI.ScriptClusterTimelineView.prototype._handleTargetPathComponentSelected): Added.
(WI.ScriptClusterTimelineView.prototype._scriptClusterViewCurrentContentViewDidChange):
(WI.ScriptClusterTimelineView.prototype.get eventsContentView): Deleted.
(WI.ScriptClusterTimelineView.prototype.get profileContentView): Deleted.
(WI.ScriptClusterTimelineView.prototype.showEvents): Deleted.
(WI.ScriptClusterTimelineView.prototype.showProfile): Deleted.
(WI.ScriptClusterTimelineView.prototype._pathComponentForContentView): Deleted.
(WI.ScriptClusterTimelineView.prototype._identifierForContentView): Deleted.
(WI.ScriptClusterTimelineView.prototype._showContentViewForIdentifier): Deleted.
(WI.ScriptClusterTimelineView.prototype._pathComponentSelected): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ScriptProfileTimelineView.js:
(WI.ScriptProfileTimelineView):
(WI.ScriptProfileTimelineView.prototype.get scrollableElements):
(WI.ScriptProfileTimelineView.prototype.closed):
(WI.ScriptProfileTimelineView.prototype.get navigationItems):
(WI.ScriptProfileTimelineView.prototype.reset):
(WI.ScriptProfileTimelineView.prototype._callingContextTreeForOrientation):
(WI.ScriptProfileTimelineView.prototype._clearFocusNodes):
(WI.ScriptProfileTimelineView.prototype._updateTargetNavigationItemDisplay): Deleted.
(WI.ScriptProfileTimelineView.prototype._displayNameForTarget): Deleted.
(WI.ScriptProfileTimelineView.prototype._handleRecordingTargetAdded): Deleted.
(WI.ScriptProfileTimelineView.prototype._populateTargetNavigationItemContextMenu): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ScriptProfileTimelineView.css: Removed.

* Source/WebInspectorUI/UserInterface/Views/ScriptDetailsTimelineView.js:
(WI.ScriptDetailsTimelineView):
(WI.ScriptDetailsTimelineView.prototype._processPendingRecords):
(WI.ScriptDetailsTimelineView.prototype._processRecord):
* Source/WebInspectorUI/UserInterface/Views/ScriptTimelineDataGridNode.js:
(WI.ScriptTimelineDataGridNode.prototype.get data):
Remove logic around and data for the `target`/&quot;Context&quot; column.

* Source/WebInspectorUI/UserInterface/Views/PathComponentIcons.css:
(.page-icon .icon): Added.
(.worker-icon .icon): Added.
(@media (prefers-color-scheme: dark) .page-icon .icon): Added.
(@media (prefers-color-scheme: dark) .worker-icon .icon): Added.
Add `WI.HierarchicalPathComponent` icons to represent &quot;Page&quot; and `Worker`.
These icons are the same as what&apos;s used in the Sources Tab to represent the main frame/page resource and `Worker` script respectively.

* Source/WebInspectorUI/UserInterface/Models/ScriptTimeline.js: Added.
(WI.ScriptTimeline.prototype.get targets):
(WI.ScriptTimeline.prototype.reset):
(WI.ScriptTimeline.prototype.callingContextTree):
(WI.ScriptTimeline.prototype.updateCallingContextTrees):
* Source/WebInspectorUI/UserInterface/Models/Timeline.js:
(WI.Timeline.create):
* Source/WebInspectorUI/UserInterface/Models/TimelineRecording.js:
(WI.TimelineRecording):
(WI.TimelineRecording.async import):
(WI.TimelineRecording.prototype.reset):
(WI.TimelineRecording.prototype.updateCallingContextTrees):
(WI.TimelineRecording.prototype.get targets): Deleted.
(WI.TimelineRecording.prototype.callingContextTree): Deleted.
* Source/WebInspectorUI/UserInterface/Controllers/TimelineManager.js:
(WI.TimelineManager.prototype.scriptProfilerTrackingCompleted):
Drive-by: move the `WI.CallingContextTree` to a newly defined `WI.ScriptTimeline` so that `WI.ScriptProfileTimelineView` can use them without also requiring the entire `WI.TimelineRecording` to be provided.

* Source/WebInspectorUI/UserInterface/Views/TimelineTabContentView.js:
(WI.TimelineTabContentView.displayNameForRecord):
Drive-by: simplify naming of `WI.TargetType.Worker` to just the last path component of the main script (or the `name` if provided) to match elsewhere in Web Inspector (e.g. when paused in a `Worker` in the Sources Tab).

* Source/WebInspectorUI/UserInterface/Base/Utilities.js:
(Map.prototype.get firstKey): Added.
* LayoutTests/inspector/unit-tests/map-utilities.html:
* LayoutTests/inspector/unit-tests/map-utilities-expected.txt:
Add a helper method to grab the first key of a `Map` similar to how `Array`/`Set` have `firstValue`.

* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Test.html:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

Canonical link: <a href="https://commits.webkit.org/291074@main">https://commits.webkit.org/291074@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fad8d32192786561fbfdd9d96449c7b1302624ec

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91952 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11483 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1028 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/96887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/42557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/11830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19969 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/96887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/42557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94953 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/11830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/83286 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/96887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/11830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/881 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/41772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/11830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98916 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/19074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/19326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79132 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/78812 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19508 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/19055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/18752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/22211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/20503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
```diff
@@ -77,6 +77,8 @@ localizedStrings["%s Event Dispatched"] = "%s Event Dispatched";
 localizedStrings["%s Prototype"] = "%s Prototype";
 /* Format string for the suggested filename when saving the content for a request local override. */
 localizedStrings["%s Request Data @ Local Override Request Content View"] = "%s Request Data";
+/* Label for JavaScript heap snapshot context name. */
```
Let's help the translators with a more descriptive label so they understand what the replaced strings represent.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Base/Utilities.js:
```diff
@@ -174,7 +182,7 @@ Object.defineProperty(WeakMap.prototype, "getOrInitialize",
 
         if (typeof initialValue === "function")
-            initialValue = initialValue();
+            initialValue = initialValue(key);
```
Nice catch! 👍

dcrousso committed:
Web Inspector: Timelines: Script: make the target into a hierarchical path component
https://bugs.webkit.org/show_bug.cgi?id=288262

Reviewed by BJ Burg.

Instead of having a `WI.DataGrid` column for the `target`/"context" in `WI.ScriptDetailsTimelineView` and a dropdown to switch the `WI.CallingContextTree` based on the selected `WI.Target` in `WI.ScriptProfileTimelineView`, the `WI.Target` should be a hierarchical path component to have similar behavior to elsewhere in Web Inspector.

This also means that distinct `WI.ScriptProfileTimelineView` and `WI.ScriptDetailsTimelineView` are created for each `WI.Target` (as needed) instead of having each hold and switch between or display all `WI.Target` in the `WI.TimelineRecording`.

Note that the `WI.HierarchicalPathComponent` for the `WI.Target` is only shown if there's more than one `WI.Target` in the `WI.TimelineRecording`.

* Source/WebInspectorUI/UserInterface/Views/ScriptClusterTimelineView.js:
(WI.ScriptClusterTimelineView):
(WI.ScriptClusterTimelineView.createPathComponent):
(WI.ScriptClusterTimelineView.prototype.reset):
(WI.ScriptClusterTimelineView.prototype.get selectionPathComponents):
(WI.ScriptClusterTimelineView.prototype.saveToCookie):
(WI.ScriptClusterTimelineView.prototype.restoreFromCookie):
(WI.ScriptClusterTimelineView.prototype.closed):
(WI.ScriptClusterTimelineView.prototype._updateCurrentContentView):
(WI.ScriptClusterTimelineView.prototype._createTargetPathComponent): Added.
(WI.ScriptClusterTimelineView.prototype._sortTargetPathComponents): Added.
(WI.ScriptClusterTimelineView.prototype._handleViewPathComponentSelected):
(WI.ScriptClusterTimelineView.prototype._handleTargetAdded): Added.
(WI.ScriptClusterTimelineView.prototype._handleTargetPathComponentSelected): Added.
(WI.ScriptClusterTimelineView.prototype._scriptClusterViewCurrentContentViewDidChange):
(WI.ScriptClusterTimelineView.prototype.get eventsContentView): Deleted.
(WI.ScriptClusterTimelineView.prototype.get profileContentView): Deleted.
(WI.ScriptClusterTimelineView.prototype.showEvents): Deleted.
(WI.ScriptClusterTimelineView.prototype.showProfile): Deleted.
(WI.ScriptClusterTimelineView.prototype._pathComponentForContentView): Deleted.
(WI.ScriptClusterTimelineView.prototype._identifierForContentView): Deleted.
(WI.ScriptClusterTimelineView.prototype._showContentViewForIdentifier): Deleted.
(WI.ScriptClusterTimelineView.prototype._pathComponentSelected): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ScriptProfileTimelineView.js:
(WI.ScriptProfileTimelineView):
(WI.ScriptProfileTimelineView.prototype.get scrollableElements):
(WI.ScriptProfileTimelineView.prototype.closed):
(WI.ScriptProfileTimelineView.prototype.get navigationItems):
(WI.ScriptProfileTimelineView.prototype.reset):
(WI.ScriptProfileTimelineView.prototype._callingContextTreeForOrientation):
(WI.ScriptProfileTimelineView.prototype._clearFocusNodes):
(WI.ScriptProfileTimelineView.prototype._updateTargetNavigationItemDisplay): Deleted.
(WI.ScriptProfileTimelineView.prototype._displayNameForTarget): Deleted.
(WI.ScriptProfileTimelineView.prototype._handleRecordingTargetAdded): Deleted.
(WI.ScriptProfileTimelineView.prototype._populateTargetNavigationItemContextMenu): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ScriptProfileTimelineView.css: Removed.

* Source/WebInspectorUI/UserInterface/Views/ScriptDetailsTimelineView.js:
(WI.ScriptDetailsTimelineView):
(WI.ScriptDetailsTimelineView.prototype._processPendingRecords):
(WI.ScriptDetailsTimelineView.prototype._processRecord):
* Source/WebInspectorUI/UserInterface/Views/ScriptTimelineDataGridNode.js:
(WI.ScriptTimelineDataGridNode.prototype.get data):
Remove logic around and data for the `target`/"Context" column.

* Source/WebInspectorUI/UserInterface/Views/PathComponentIcons.css:
(.page-icon .icon): Added.
(.worker-icon .icon): Added.
(@media (prefers-color-scheme: dark) .page-icon .icon): Added.
(@media (prefers-color-scheme: dark) .worker-icon .icon): Added.
Add `WI.HierarchicalPathComponent` icons to represent "Page" and `Worker`.
These icons are the same as what's used in the Sources Tab to represent the main frame/page resource and `Worker` script respectively.

* Source/WebInspectorUI/UserInterface/Models/ScriptTimeline.js: Added.
(WI.ScriptTimeline.prototype.get targets):
(WI.ScriptTimeline.prototype.reset):
(WI.ScriptTimeline.prototype.callingContextTree):
(WI.ScriptTimeline.prototype.updateCallingContextTrees):
* Source/WebInspectorUI/UserInterface/Models/Timeline.js:
(WI.Timeline.create):
* Source/WebInspectorUI/UserInterface/Models/TimelineRecording.js:
(WI.TimelineRecording):
(WI.TimelineRecording.async import):
(WI.TimelineRecording.prototype.reset):
(WI.TimelineRecording.prototype.updateCallingContextTrees):
(WI.TimelineRecording.prototype.get targets): Deleted.
(WI.TimelineRecording.prototype.callingContextTree): Deleted.
* Source/WebInspectorUI/UserInterface/Controllers/TimelineManager.js:
(WI.TimelineManager.prototype.scriptProfilerTrackingCompleted):
Drive-by: move the `WI.CallingContextTree` to a newly defined `WI.ScriptTimeline` so that `WI.ScriptProfileTimelineView` can use them without also requiring the entire `WI.TimelineRecording` to be provided.

* Source/WebInspectorUI/UserInterface/Views/TimelineTabContentView.js:
(WI.TimelineTabContentView.displayNameForRecord):
Drive-by: simplify naming of `WI.TargetType.Worker` to just the last path component of the main script (or the `name` if provided) to match elsewhere in Web Inspector (e.g. when paused in a `Worker` in the Sources Tab).

* Source/WebInspectorUI/UserInterface/Base/Utilities.js:
(Map.prototype.get firstKey): Added.
* LayoutTests/inspector/unit-tests/map-utilities.html:
* LayoutTests/inspector/unit-tests/map-utilities-expected.txt:
Add a helper method to grab the first key of a `Map` similar to how `Array`/`Set` have `firstValue`.

* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Test.html:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

Canonical link: https://commits.webkit.org/291074@main
